build.sh

```diff
@@ -68,33 +65,64 @@ function docker_build_with_version {
                "awk '/Successfully built/{print \$NF}'" \
                IMAGE_ID
 
-  name=$(docker inspect -f "{{.Config.Labels.name}}" $IMAGE_ID)
 
-  IMAGE_NAME=$name
-  if [ -n "${TEST_MODE}" ]; then
-    IMAGE_NAME+="-candidate"
+  if test -f .image-id.raw; then
+      local previous_id=`cat .image-id.raw`
+      if test "$IMAGE_ID" != "$previous_id"; then
+          # Try to do "best-effort" cleanup the previous one.
+          docker rmi "$previous_id" || :
+      fi
   fi
-  echo "-> Image ${IMAGE_ID} tagged as ${IMAGE_NAME}"
-  docker tag $IMAGE_ID $IMAGE_NAME
+  echo "$IMAGE_ID" > .image-id.raw
 
-  if [[ "${SKIP_SQUASH}" != "1" ]]; then
-    docker tag $IMAGE_ID "${IMAGE_NAME}-unsquashed"
-    squash "${dockerfile}"
-  fi
-  # Narrow by repo:tag first and then grep out the exact match
-  docker images "${IMAGE_NAME}:latest" --format="{{.Repository}} {{.ID}}" | grep "^${IMAGE_NAME}" | awk '{print $2}' >.image-id
+  squash "${dockerfile}" "$IMAGE_ID"
+  echo "$IMAGE_ID" > .image-id
 }
 
+# squash DOCKERFILE
+# -----------------
 # Install the docker squashing tool[1] and squash the result image
+# when necessary.
 # [1] https://github.com/goldmann/docker-squash
-function squash {
+# Reads:
+#   $IMAGE_ID
+# Sets:
+#   $IMAGE_ID
+squash ()
+{
+  local base squashed_from squashed= unsquashed=$IMAGE_ID
+  test "$SKIP_SQUASH" = 1 && return 0
+
   # FIXME: We have to use the exact versions here to avoid Docker client
   #        compatibility issues
   local squash_version=1.0.5
```

Comment:
@omron93 wrote:
> I vote for not checking for exact docker_squash version.

+1.  I was not brave enough to drop that ... since this hack was important enough to be fixed many, many times ...  So the question is whether we really care about the version or not.  Maybe there should be `version >= 1.0.5`, but IMO this deserves separate PR and discussion.

> Imagine a situation when we decides to change this version. We/user will have to update docker_squash on testing machine to make similar PR tests to pass.

Right, but IMO it is not that huge issue, since we maintain the CI.  I don't think user will be somewhat bothered by this.

> But since other images won't use newest version of container-common-scripts its PRs will fail.

Agreed. IMO in such case it would be the right time to update `common` in the remote repos; and yes -> I don't claim that such hacks are not expensive.

So, since the version check was already merged and is not the part of this PR, what about to remove that later (and let others ack this?)?